consensus/istanbul/core/commit.go

```diff
@@ -39,15 +39,40 @@ func (c *core) sendCommitForOldBlock(view *istanbul.View, digest common.Hash) {
 	c.broadcastCommit(sub)
 }
 
-func (c *core) generateCommittedSeal(sub *istanbul.Subject) ([]byte, error) {
+func (c *core) generateCommittedSeal(sub *istanbul.Subject) (blscrypto.SerializedSignature, error) {
 	seal := PrepareCommittedSeal(sub.Digest, sub.View.Round)
 	committedSeal, err := c.backend.SignBlockHeader(seal)
 	if err != nil {
-		return nil, err
+		return blscrypto.SerializedSignature{}, err
 	}
 	return committedSeal, nil
 }
 
+func (c *core) generateEpochSeal() (blscrypto.SerializedSignature, error) {
+  //TODO(kobi): is this the correct handling?
+  if c.current.Proposal() == nil {
+    return blscrypto.SerializedSignature{}, nil
+  }
+	currentBlockNumber := c.current.Proposal().Number().Uint64()
+	if !istanbul.IsLastBlockOfEpoch(currentBlockNumber, c.config.Epoch) {
+    return blscrypto.SerializedSignature{}, nil
+  }
+  valSet := c.current.ValidatorSet()
+  blsPubKeys := []blscrypto.SerializedPublicKey{}
+  for _, v := range valSet.List() {
+    blsPubKeys = append(blsPubKeys, v.BLSPublicKey())
+  }
+  epochData, err := blscrypto.EncodeEpochSnarkData(blsPubKeys, uint32(valSet.F()+1), uint16(currentBlockNumber/c.config.Epoch))
```

Comment:
You can use this function to calculate the current epoch number:  https://github.com/celo-org/celo-blockchain/blob/master/consensus/istanbul/utils.go#L90

Reply:
Ah, useful!